### ISSUE TITLE
Great work.  There is something I hope dear authors can cast light on for me.

### ISSUE THREAD

**Issue by onbigion13** [sangwoohwang/SpikedAttention#1]
In you papser, you mentioned
“In previous works (12; 14), logical AND operations are performed between two “rate-coded" spike trains, e.g., $S_{Q(i,k)}$, and $S_{K(j,k)}$.”
Here reference 12 is the research of "Spikformer".

Their implementation of SSA is as following:
```python3
class SSA(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0., sr_ratio=1):
        super().__init__()
        assert dim % num_heads == 0, f"dim {dim} should be divided by num_heads {num_heads}."
        self.dim = dim
        self.num_heads = num_heads
        self.scale = 0.125
        self.q_conv = nn.Conv1d(dim, dim, kernel_size=1, stride=1, bias=False)
        self.q_bn = nn.BatchNorm1d(dim)
        self.q_lif = MultiStepLIFNode(tau=2.0, detach_reset=True, backend='cupy')

        self.k_conv = nn.Conv1d(dim, dim, kernel_size=1, stride=1, bias=False)
        self.k_bn = nn.BatchNorm1d(dim)
        self.k_lif = MultiStepLIFNode(tau=2.0, detach_reset=True, backend='cupy')

        self.v_conv = nn.Conv1d(dim, dim, kernel_size=1, stride=1, bias=False)
        self.v_bn = nn.BatchNorm1d(dim)
        self.v_lif = MultiStepLIFNode(tau=2.0, detach_reset=True, backend='cupy')
        self.attn_lif = MultiStepLIFNode(tau=2.0, v_threshold=0.5, detach_reset=True, backend='cupy')

        self.proj_conv = nn.Conv1d(dim, dim, kernel_size=1, stride=1)
        self.proj_bn = nn.BatchNorm1d(dim)
        self.proj_lif = MultiStepLIFNode(tau=2.0, detach_reset=True, backend='cupy')

    def forward(self, x, res_attn):
        T, B, C, H, W = x.shape
        x = x.flatten(3)
        T, B, C, N = x.shape
        x_for_qkv = x.flatten(0, 1)
        q_conv_out = self.q_conv(x_for_qkv)
        q_conv_out = self.q_bn(q_conv_out).reshape(T, B, C, N).contiguous()
        q_conv_out = self.q_lif(q_conv_out)
        q = q_conv_out.transpose(-1, -2).reshape(T, B, N, self.num_heads, C // self.num_heads).permute(0, 1, 3, 2,
                                                                                                       4).contiguous()

        k_conv_out = self.k_conv(x_for_qkv)
        k_conv_out = self.k_bn(k_conv_out).reshape(T, B, C, N).contiguous()
        k_conv_out = self.k_lif(k_conv_out)
        k = k_conv_out.transpose(-1, -2).reshape(T, B, N, self.num_heads, C // self.num_heads).permute(0, 1, 3, 2,
                                                                                                       4).contiguous()

        v_conv_out = self.v_conv(x_for_qkv)
        v_conv_out = self.v_bn(v_conv_out).reshape(T, B, C, N).contiguous()
        v_conv_out = self.v_lif(v_conv_out)
        v = v_conv_out.transpose(-1, -2).reshape(T, B, N, self.num_heads, C // self.num_heads).permute(0, 1, 3, 2,
                                                                                                       4).contiguous()

        x = k.transpose(-2, -1) @ v
        x = (q @ x) * self.scale

        x = x.transpose(3, 4).reshape(T, B, C, N).contiguous()
        x = self.attn_lif(x)
        x = x.flatten(0, 1)
        x = self.proj_lif(self.proj_bn(self.proj_conv(x)).reshape(T, B, C, H, W))
        return x, v
```
I did not find any implementation of logic And operator in their implementation.

I am doing a research , and need some lightning implementation of Matmul operator. I accidentally read your research, fortunately find that you mentioned that there are some researches implementing certain operators in SSA as logic AND, if you know any other works  like this, please share with me. 

I really admire you marvellous research, and appreciate your help, if your would like to share with me. 

Best regards.
Gt.

**Comment by onbigion13** [comment]
Also I check the implementation in the official repo of  "Mask Spiking Transformer".

Their spiking MLP is implementation as following:
```python

class SpikingMlp(nn.Module):
    def __init__(self, in_features, hidden_features=None, out_features=None, drop=0., masking_ratio=0.,
                 dataset='imagenet'):
        super().__init__()
        out_features = out_features or in_features
        hidden_features = hidden_features or in_features
        self.dataset = dataset
        self.fc1 = nn.Linear(in_features, hidden_features)
        if dataset == 'imagenet':
            self.norm1 = nn.BatchNorm1d(hidden_features)
        self.fc1_if = nn.ReLU()
        self.fc2 = nn.Linear(hidden_features, out_features)
        if dataset == 'imagenet':
            self.norm2 = nn.BatchNorm1d(out_features)
        self.fc2_if = nn.ReLU()
        self.drop = nn.Dropout(drop)
        self.masking_ratio = masking_ratio
        self.out_features = out_features

    def forward(self, x):
        x = self.fc1(x)
        if self.dataset == 'imagenet':
            x = self.norm1(x.transpose(1, 2).contiguous()).transpose(1, 2).contiguous()
        x = self.fc1_if(x)
        x = self.drop(x)
        x = self.fc2(x)
        if self.dataset == 'imagenet':
            x = self.norm2(x.transpose(1, 2).contiguous()).transpose(1, 2).contiguous()
        x = self.fc2_if(x)
        x = self.drop(x)
        return x

```

They also built the layer with existing floating `nn.Linear`;  I failed to find any logic operator in their implementation.
